src/Glimpse.Server.Web/Storage/InMemoryStorage.cs

```diff
@@ -8,65 +7,200 @@ namespace Glimpse.Server.Web
+
+            /// <summary>
+            /// Indices associated with this Request
+            /// </summary>
+            public RequestIndices Indices { get; set; }
+        }
+
         public const int RequestsPerPage = 50;
-        public const int MaxRequests = 500; // TODO: Make this configurable
+        public const int DefaultMaxRequests = 500;
 
-        private readonly List<IMessage> _messages;
-        private readonly ConcurrentDictionary<Guid, RequestIndices> _indices;
+        /// <summary>
+        /// Primary storage for Messages. 
+        /// </summary>
+        private readonly Dictionary<Guid, RequestInfo> _requestTable;
 
-        private static readonly object _locker = new object(); // HACK :(
+        /// <summary>
+        /// Currently active requests, ordered by incoming messages for the request.  The first entry in the list will be
+        /// the most recent request to receive a message, the last entry the oldest request to receive a message.
+        /// </summary>
+        private readonly LinkedList<Guid> _activeRequests;
 
-        public InMemoryStorage()
+        /// <summary>
+        /// Maximum number of requests to store for this request.
+        /// </summary>
+        private readonly int _maxRequests;
+
+
+        /// <summary>
+        /// Lock to synchronize access to this data structure. 
+        /// </summary>
+        // TODO:  Consider changing to a multi-reader/single-writer lock for increased read throughput
+        private static readonly object _locker = new object();
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="maxRequests">Max number of requests to store.  OPtional.  Defaults to DefaultMaxRequests.</param>
+        public InMemoryStorage(int maxRequests = DefaultMaxRequests)
         {
-            _messages = new List<IMessage>();
-            _indices = new ConcurrentDictionary<Guid, RequestIndices>();
+            _requestTable = new Dictionary<Guid, RequestInfo>();
+            _activeRequests = new LinkedList<Guid>();
+            _maxRequests = maxRequests;
         }
 
+        /// <summary>
+        /// Persist the given message.
+        /// </summary>
+        /// <param name="message">The message to store.</param>
         public void Persist(IMessage message)
         {
-            _messages.Add(message);
+            Guid requestId = message.Context.Id;
+
+            lock (_locker)
+            {
+                RequestInfo ri;
+                if (!_requestTable.TryGetValue(requestId, out ri))
+                {
+                    LinkedListNode<Guid> llNode = _activeRequests.AddFirst(requestId);
+                    ri = new RequestInfo(llNode);
+                    _requestTable.Add(requestId, ri);
+                }
+                else
+                {
+                    // move this request to the head of the "active list"
+                    if (ri.RequestIdNode.Previous != null)
```

Comment:
Fixed.

From: Anthony van der Hoorn [mailto:notifications@github.com]
Sent: Friday, September 25, 2015 9:55 AM
To: Glimpse/Glimpse.Prototype <Glimpse.Prototype@noreply.github.com>
Cc: Mike Kaufman <Mike.Kaufman@microsoft.com>
Subject: Re: [Glimpse.Prototype] 55 in memory storage datastructures (#62)


In src/Glimpse.Server.Web/Storage/InMemoryStorage.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fGlimpse%2fGlimpse.Prototype%2fpull%2f62%23discussion_r40450751&data=01%7c01%7cmike.kaufman%40microsoft.com%7c19548a27a2fe4d21c1e208d2c5ca0558%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=KIjUfxDV7zCM%2bN1vpxlq%2bD4uufTllQwh43WceHRfyw4%3d>:

> -            _messages.Add(message);

> +            Guid requestId = message.Context.Id;

> +

> +            lock (_locker)

> +            {

> +                RequestInfo ri;

> +                if (!_requestTable.TryGetValue(requestId, out ri))

> +                {

> +                    LinkedListNode<Guid> llNode = _activeRequests.AddFirst(requestId);

> +                    ri = new RequestInfo(llNode);

> +                    _requestTable.Add(requestId, ri);

> +                }

> +                else

> +                {

> +                    // move this request to the head of the "active list"

> +                    if (ri.RequestIdNode.Previous != null)

Nit... This could probably be an else if?

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fGlimpse%2fGlimpse.Prototype%2fpull%2f62%2ffiles%23r40450751&data=01%7c01%7cmike.kaufman%40microsoft.com%7c19548a27a2fe4d21c1e208d2c5ca0558%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=IIqD8eTX%2fG8gReauBtW2YzJyUBUcoLg%2biwpgcyCIUeM%3d>.